civitai-downloader-v2/tests/unit/test_database_optimization.py

```
@@ -0,0 +1,440 @@
+            
+            # Test batch insert is faster per item
+            individual_rate = 10 / individual_time if individual_time > 0 else float('inf')
+            batch_rate = len(remaining_models) / batch_time if batch_time > 0 else float('inf')
+            
+            # Batch should be at least as fast as individual (usually much faster)
+            assert batch_rate >= individual_rate * 0.5, "Batch insert should be efficient"
+            
+            # Test transaction rollback on error
+            invalid_models = [{'id': 'invalid', 'name': 'Invalid Model'}]  # Invalid ID type
+            
+            with pytest.raises(Exception):
```

Ruff (0.12.2)

Do not assert blind exception: `Exception`

(B017)